src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs

```diff
@@ -190,39 +125,23 @@ private Task ExecutePageAsync()
 
                 _page = (Page)CacheEntry.PageFactory(_pageContext, _viewContext);
 
-                pageResult.Page = _page;
-                pageResult.ViewData = pageResult.ViewData ?? _pageContext.ViewData;
+                if (_actionDescriptor.ModelTypeInfo == _actionDescriptor.PageTypeInfo)
```

Comment:
No, there are two cases where `!HasPageModel` is false. You can have handler methods on the page it self and still use `@model`.